### PR TITLE
Rename the field 'provctx and data' to 'algctx' inside some EVP objects.

### LINKED ISSUE #14284: Change provctx fields in EVP_xxx_CTX structures to data.
Most `EVP_xxx_CTX` structures keep the provider side context pointer in a field called `data`.  The rest (EVP_MD_CTX, EVP_CIPHER_CTX, ...) use the name `provctx` which is confusing.

This was first raised [here](https://github.com/openssl/openssl/pull/14240#discussion_r578879244).  Again in more details [here](https://github.com/openssl/openssl/pull/14240#discussion_r580695418).

### PR DESCRIPTION
**Pull request by slontis**
Fixes #14284

This was a bit confusing since some objects need an actual provider
context to be passed (e.g. asym_cipher_gettable_ctx_params)
This 'data' field is an opaque ptr returned from new_ctx() operations. The
ptr points to data created inside a providers implementation of an
algorithm. This ptr is then passed to the provider algorithm for all
subsequent operations (such as init(), update(), and free()).

##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

File: include/crypto/evp.h [hunk 1]
```diff
@@ -58,21 +58,21 @@ struct evp_pkey_ctx_st {
 
         struct {
             EVP_KEYEXCH *exchange;
-            void *exchprovctx;
+            void *data;
```
Comment (slontis) [inline include/crypto/evp.h:61]: Should we move this 'data' outside the struct?

File: crypto/evp/evp_local.h [hunk 2]
```diff
@@ -27,8 +25,11 @@ struct evp_md_ctx_st {
     /* Update function: usually copied from EVP_MD */
     int (*update) (EVP_MD_CTX *ctx, const void *data, size_t count);
 
-    /* Provider ctx */
-    void *provctx;
+    /*
+     * Opaque data returned from a providers digest algorithm implementation
+     * OSSL_FUNC_digest_newctx()
+     */
+    void *data;
```
Comment (levitte) [inline crypto/evp/evp_local.h:32]: How about `mdctx`?

File: crypto/evp/evp_local.h [hunk 3]
```diff
@@ -51,8 +52,11 @@ struct evp_cipher_ctx_st {
     int block_mask;
     unsigned char final[EVP_MAX_BLOCK_LENGTH]; /* possible final block */
 
-    /* Provider ctx */
-    void *provctx;
+    /*
+     * Opaque data returned from a providers cipher algorithm implementation
+     * OSSL_FUNC_cipher_newctx()
+     */
+    void *data;
```
Comment (levitte) [inline crypto/evp/evp_local.h:59]: How about `ciphctx`?

File: include/crypto/evp.h (hunk 1 above)
Comment (levitte) [inline include/crypto/evp.h:61]: How about `exchctx`?

File: include/crypto/evp.h [hunk 4]
```diff
@@ -58,21 +58,21 @@ struct evp_pkey_ctx_st {
...
         } kex;
 
         struct {
             EVP_SIGNATURE *signature;
-            void *sigprovctx;
+            void *data;
```
Comment (levitte) [inline include/crypto/evp.h:66]: How about `sigctx`?

File: include/crypto/evp.h [hunk 5]
```diff
@@ -58,21 +58,21 @@ struct evp_pkey_ctx_st {
...
         } sig;
 
         struct {
             EVP_ASYM_CIPHER *cipher;
-            void *ciphprovctx;
+            void *data;
```
Comment (levitte) [inline include/crypto/evp.h:71]: How about `ciphctx`?

File: include/crypto/evp.h [hunk 6]
```diff
@@ -58,21 +58,21 @@ struct evp_pkey_ctx_st {
...
         } ciph;
         struct {
             EVP_KEM *kem;
-            void *kemprovctx;
+            void *data;
```
Comment (levitte) [inline include/crypto/evp.h:75]: How about `kemctx`?

**Review by levitte — CHANGES_REQUESTED** [review]
`data` is too nondescript in my view, could just as well be `xyzzy`.  I understand wanting to avoid the confusion with the provider context, though.
So here's another idea.

File: include/crypto/evp.h (hunk 1 above)
Comment (levitte) [reply include/crypto/evp.h:61]: No, let's leave it visually encapsulated

File: crypto/evp/evp_local.h (hunk 3 above)
Comment (slontis) [reply crypto/evp/evp_local.h:59]: So then we end up with ciphctx sitting inside evp_cipher_ctx_st which is still kind of confusing..
I think it was provctx to indicate that it was a provider side ctx..
Should we go with something even more longwinded like provciphctx ?

Comment (levitte) [reply crypto/evp/evp_local.h:59]: If you prefer the longwinded variant, go for it.  But then, how about `ciphprovctx` to match the already existing names in EVP_PKEY_CTX (less renaming for you to do as well :wink:)

Comment (slontis) [reply crypto/evp/evp_local.h:59]: But then we are kind of back where we started with a provctx - which is not a good place to be.

Comment (levitte) [reply crypto/evp/evp_local.h:59]: Ah, so you mean that `provciphctx` seems clearer than `ciphprovctx`?  I didn't quite catch on to that...  it doesn't make a difference for me, but if it does for you, then do it!

Comment (slontis) [reply crypto/evp/evp_local.h:59]: Anyone else have an opinion?

**Comment by levitte** [comment]
Side note: I *like* what the checksums github action does.  Nice work @t8m!

File: crypto/evp/evp_local.h (hunk 3 above)
Comment (kaduk) [reply crypto/evp/evp_local.h:59]: The bikeshed shall be painted lilac!
Erm, I mean, `provciphctx` seems more clear to me than `ciphprovctx`.

Comment (paulidale) [reply crypto/evp/evp_local.h:59]: I'd be perfectly fine with `provctx`.  The point of the issue was that the names were inconsistent `data` in some places and `provctx` in others.  So long as they are bought in line, I'm not fussed which way.

I'm not sure including `ciph` or similar is adding anything here.  In fact it reduces readability IMO.

Comment (slontis) [reply crypto/evp/evp_local.h:59]: Err that was one of the points, The other point was that it is not the provider context. It is an allocated blob of data (some form of ctx) that resides in an algorithm implementation (which sits inside a provider). provctx is too generic a term for use here.

Comment (paulidale) [reply crypto/evp/evp_local.h:59]: `provdata` perhaps?

Comment (levitte) [reply crypto/evp/evp_local.h:59]: I would rather avoid `data` in any form for these contexts, as it suggests some form of payload in my mind. See `keydata`

Comment (paulidale) [reply crypto/evp/evp_local.h:59]: `blob` then.  Or `provblob`.

Comment (slontis) [reply crypto/evp/evp_local.h:59]: algctx or provalgctx?

Comment (slontis) [reply crypto/evp/evp_local.h:59]: The thing this maps to is PROV_AES_CTX for example.

Comment (levitte) [reply crypto/evp/evp_local.h:59]: If you want the One Generic Name, `algctx` gets my vote

**Comment by slontis** [comment]
Changed all of them to use algctx

File: crypto/evp/evp_local.h [hunk 7]
```diff
@@ -27,8 +25,11 @@ struct evp_md_ctx_st {
     /* Update function: usually copied from EVP_MD */
     int (*update) (EVP_MD_CTX *ctx, const void *data, size_t count);
 
-    /* Provider ctx */
-    void *provctx;
+    /*
+     * Opaque ctx returned from a providers digest algorithm implementation
+     * OSSL_FUNC_digest_newctx()
+     */
```
Comment (levitte) [inline crypto/evp/evp_local.h:31]: This comment seems a bit bloated at this point.  How about `/* Algorithm-specific ctx */`?

File: crypto/evp/evp_local.h [hunk 8]
```diff
@@ -51,24 +52,27 @@ struct evp_cipher_ctx_st {
     int block_mask;
     unsigned char final[EVP_MAX_BLOCK_LENGTH]; /* possible final block */
 
-    /* Provider ctx */
-    void *provctx;
+    /*
+     * Opaque ctx returned from a providers cipher algorithm implementation
+     * OSSL_FUNC_cipher_newctx()
+     */
```
Comment (levitte) [inline crypto/evp/evp_local.h:58]: This comment seems a bit bloated at this point.  How about `/* Algorithm-specific ctx */`?

**Review by levitte — CHANGES_REQUESTED** [review]
Other than this, I'm happy, and especially that you fixed the MAC code accordingly!

File: crypto/evp/evp_local.h (hunk 7 above)
Comment (paulidale) [reply crypto/evp/evp_local.h:31]: I prefer the longer comment.

**Review by paulidale — APPROVED** [review]

**Review by t8m — APPROVED** [review]
Nice cleanup!

File: crypto/evp/evp_local.h (hunk 7 above)
Comment (levitte) [reply crypto/evp/evp_local.h:31]: So why only here and not *everywhere where we have the algctx field?

**Comment by slontis** [comment]
Changed the comments..

File: include/crypto/evp.h [hunk 9]
```diff
@@ -58,20 +58,36 @@ struct evp_pkey_ctx_st {
...
             void *algctx;
         } kex;
 
         struct {
+            /*
+             * Opaque ctx returned from a providers signature algorithm
+             * implementation OSSL_FUNC_signature_newctx()
+             */
             EVP_SIGNATURE *signature;
             void *algctx;
```
Comment (levitte) [inline include/crypto/evp.h:74]: Comment slightly misplaced...

**Review by levitte — CHANGES_REQUESTED** [review]

File: crypto/evp/evp_local.h (hunk 7 above)
Comment (levitte) [reply crypto/evp/evp_local.h:31]: Change to make them all more verbose noted.

I find it amazing, though, that extended verbosity is chosen for this item specifically.  I kinda understand it considering that there seems to be a lot of confusion...  but in a year, will we still feel the same?

Comment (levitte) [reply crypto/evp/evp_local.h:31]: Anyway, I'm not going to block on terms of verbosity, just wanted to express my surprise over this choice, *this time*

Comment (slontis) [reply crypto/evp/evp_local.h:31]: This may be trying to be kind to others trying to understand what it is.

**Review by levitte — APPROVED** [review]

**Review by t8m — APPROVED** [review]

**Comment by paulidale** [comment]
Merged to master.

**Comment by slontis** [comment]
Thanks Pauli..

**Pull request closed without merging**